src/neorados/cls/common.h

```diff
@@ -46,6 +47,38 @@
 /// Ideally, these functions will help.
 
 namespace neorados::cls {
+namespace detail {
+template<typename...>
+struct return_sig;
+
+template<typename T>
+struct return_sig<T> {
+using type = void(boost::system::error_code, T);
+};
+template<typename ...Ts>
+struct return_sig<std::tuple<Ts...>> {
+  using type = void(boost::system::error_code, Ts...);
+};
+
+template<typename... Ts>
+using return_sig_t = typename return_sig<Ts...>::type;
+
+template<typename T>
+auto maybecat(boost::system::error_code ec,
+	      T&& t)
+{
+  return std::make_tuple(ec, std::forward<T>(t));
+}
+
+template<typename ...Ts>
+auto maybecat(boost::system::error_code ec,
```

Comment:
looks like a homebrew std::expected in C++23 or boost::outcome. If we don't have the luxury of using the former, probably we can use the latter.

Reply:
We have a version of `expected` in `src/include` (Though we may not need it since Boost.System seems to have added one of their own.)

But this isn't that. This is a result of Asio completion handlers having a call signature like `boost::sysem::error_code, stuff, other stuff, ...`

`neorados::cls::exec` is a convenience function that packages up the call and encode/decode stuff, you just have to provide a function that pulls the value you're interested in out of the response structure. `maybecat` handles the problem of having multiple items you want to pull out (and returning them as a tuple.)

That way we end up with our handler having the expected signature of `void(error_code, T1, T2, T3..)` instead of `void(error_code, std::tuple<T1, T2, T3…>)`

Reply:
I see, so they are for different use cases. Thank you for the detailed explanation!